src/main/java/org/javarosa/core/model/instance/ExternalDataInstance.java

```diff
@@ -50,39 +53,36 @@ public ExternalDataInstance(ExternalDataInstance instance) {
         super(instance.getInstanceId());
         this.reference = instance.getReference();
         this.base = instance.getBase();
-        this.root = instance.getRoot();
+        //Copy constructor avoids check.
+        this.root = instance.root;
         this.mCacheHost = instance.getCacheHost();
-        this.remoteUrl = instance.getRemoteUrl();
+        this.source = instance.getSource();
         useCaseTemplate = CaseInstanceTreeElement.MODEL_NAME.equals(instanceid);
     }
 
     private ExternalDataInstance(String reference, String instanceId,
```

Comment:
should `instanceId` also come from `source`?

Reply:
There are a few of places where `instanceId` get's passed as arguments on methods for objects that have an `instanceId` field which I find confusing. e.g. `ExternalDataInstance.initialize`

Reply:
I 100% agree about `instanceId`. I didn't get a chance to go back and refactor around, but I think I accidentally added at least one new unnecessary instance of it, and its role is generally quite obtuse. 

One thing Shubham and I chatted about this morning was refactoring more substantially to replace the use of ExternalDataInstance with the new `Source` object in the `StackFrameStep` itself, which I *think* would fix a lot of the callstack ordering confusion here.